[NG] Issue #335 - Create Masked Input Component

LokeshN opened the pull request:
Added a new directive "clr-mask" for masking the data entered in input field.

Supports the following pattern,
# or 9 - for numbers
A - for a-z and A-Z
* - for a-z,A-Z,0-9

Kindly let me know in case of any issue.

youdz commented on src/clarity-angular/input-fields/providers/data-mask.ts:
```diff
@@ -0,0 +1,72 @@
+                continue;
+            } else {
+                //stop the processing, if the regex does not match
+                break;
```
You are missing a graceful error display or at least some feedback to the user when they're typing something wrong (whether it's a wrong character, or continue typing once max length has been reached, ...). Just ignoring use input makes them wonder if the app is not frozen or bugged.

LokeshN replied:
Added an output (error) event to capture the error.

LokeshN commented:
@youdz ,
Thanks for the comments.
Now I have modified the directive to implement ControlValueAccessor and using the input field event listener to listen to the change in the field value.
I am also emitting an error event when there is an error in the input text, which can be caught and users can perform action accordingly.
Also the users can configiure the directive clr-mask,
<input clr-mask="99?9?.99?9?.99?9?" (error)="onError($event)" [definition]="customMask"/>
While clr-mask is the mask configuration, error is the event which will give the error message and 
definition is the input through the user can configure the CustomMask definition.
Also added unit tests for the component itself. 
Kindly review and let me know in case of any issue.

Thanks

LokeshN commented on src/clarity-angular/input-fields/input-mask.ts:
```diff
@@ -0,0 +1,146 @@
+    providers: [DataMask, {provide: MASK_PATTERN, useValue: maskDefinitionPattern}]
+})
+export class InputMask {
+    @Input() clrInputMask: string;
```
Done

LokeshN commented on src/clarity-angular/input-fields/input-mask.ts:
```diff
@@ -0,0 +1,146 @@
+        if (this.isDeleteOrBackspaceEvent($event) && this.isHavingMinLength()) {
+            this.applyMask();
+            return false;
+        }
```
Added unit tests.

LokeshN commented on src/clarity-angular/input-fields/input-mask.ts:
```diff
@@ -0,0 +1,146 @@
+    private applyMask($event?: any): void {
+        let maskedVal: string = this.getMaskedValue($event);
+        this.setMaskedValue(maskedVal);
+        this.resetCaretPosition(maskedVal.length);
```
Fixed it.

LokeshN commented on src/clarity-angular/input-fields/input-mask.ts:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    private isPasteEvent($event: any): boolean {
+        if ($event.ctrlKey && $event.keyCode === 86) {
```
Not using keypress event now, using the input event listener.

LokeshN commented on src/clarity-angular/input-fields/input-mask.ts:
```diff
@@ -0,0 +1,146 @@
+
+    private resetCaretPosition(pos: number): void {
+        (<HTMLInputElement>this._elementRef.nativeElement)
+            .setSelectionRange(pos, pos);
```
Fixed it

LokeshN commented on src/clarity-angular/input-fields/all.spec.ts:
```diff
@@ -0,0 +1,18 @@
+describe("Input Fields", function() {
+    describe("Providers", function() {
+        DataMaskProviderSpecs();
+    });
```
Added the unit tests.

LokeshN commented on src/clarity-angular/input-fields/input-mask.ts:
```diff
@@ -0,0 +1,146 @@
+                    (keyCode > 95 && keyCode < 112)  || // numpad keys
+                    (keyCode > 185 && keyCode < 193) || // ;=,-./` (in order)
+                    (keyCode > 218 && keyCode < 223) ; //delete,backspace
+        return valid;
```
Fixed.

LokeshN commented on src/clarity-angular/input-fields/interfaces/mask-definition-pattern.ts:
```diff
@@ -0,0 +1,28 @@
+ */
+export const maskDefinitionPattern: IMaskDefinitionPattern = {
+    maskPattern: {
+        "#" : /^[0-9]+$/, // to indicate only numbers
```
Fixed.

LokeshN commented on src/clarity-angular/input-fields/interfaces/mask-definition-pattern.ts:
```diff
@@ -0,0 +1,28 @@
+        "9" : /^[0-9]+$/, // to indicate only numbers
+        "A" : /^[a-zA-Z]+$/, // to indicate any character a-z and A-Z
+        "*" : /^[a-zA-z0-9]+$/ // to indicate any character a-z, A-Z and 0-9
+    }
```
Using the patterns, but provided a way now for the users to extend the functionality using (definition) input.

LokeshN commented on src/clarity-angular/input-fields/providers/data-mask.spec.ts:
```diff
@@ -0,0 +1,64 @@
+            let maskedVal = this.dataMask.applyMask("999.999.999.999",
+                "255255255255");
+            expect(maskedVal).toEqual("255.255.255.255");
+        });
```
Added a way to provide a optional char using '?'.

LokeshN commented:
@youdz , Sorry I force pushed my changes previously, so lot of code became outdated but also I have changed it completely. Kindly let me know in case you see some issues in the code.

Thanks,
Lokesh

jaffoneh commented:
With a big pull request such as this one, the expectation is that we need to discuss this a bit more including having a design and implementation spec. This change includes hundreds of lines of codes and with a force push especially, it is hard to continue to review it without having a spec that includes both the design and implementation details that we should discuss before pushing as much code.

@LokeshN could you provide a bit more details on the assumptions you're making, the API you're building for this component, as well as the design expectations of the components and we can work together on discussing that before we discuss the code itself so we have a high-level agreement considering this is a component not just a small contribution.

LokeshN commented:
@jaffoneh, @youdz  sorry I did the force update on the PR, but I have almost rewritten everything after youdz's comments.
I have created a design doc in Google Docs, https://docs.google.com/document/d/1DN5XUGxquubPCoheLVHS9Rl2fSssejK67oeC6zAPfBs/edit

Let me know in case you need more details.

jaffoneh commented:
Thanks, @LokeshN. @youdz: could you take a look and provide comments on the Google Doc attached above to drive the engineering and design requirements?

LokeshN commented:
@youdz , @jaffoneh Kindly let me know if you are facing some issues in viewing the document.

Thanks,
Lokesh

LokeshN commented:
@youdz , @jaffoneh, Kindly let me know in case there are any issues in the design and code.
Thanks,
Lokesh

jaffoneh commented:
@LokeshN will do once we take a look. Currently, masked input is not at the top of our priorities so apologizes for the delay but this might be delayed for a bit as we finalize a couple of other components first. Let me know if this is required as a priority for you to work together on seeing how we prioritize it.
